pkg/stream/stream.go

```diff
@@ -0,0 +1,241 @@
+		storeRequests: make(chan structs.BlockAndTrace, cfg.StreamQueueSize),
+		Config:        cfg,
+		Logger:        cfg.Logger.WithField("relay-service", "stream").WithField("type", "redis"),
+	}
+
+	s.initMetrics()
+
+	return &s
+}
+
+func (s *RedisStream) RunSubscriberParallel(ctx context.Context, ds Datastore, num uint) error {
+	blocks := s.Pubsub.Subscribe(ctx, s.Config.PubsubTopic)
+
+	for i := uint(0); i < num; i++ {
+		go s.RunSubscriber(ctx, ds, blocks)
+	}
+
+	return nil
+}
+
+func (s *RedisStream) RunSubscriber(ctx context.Context, ds Datastore, blocks chan []byte) error {
+	sBlock := StreamBlock{}
+
+	for rawSBlock := range blocks {
+		if err := proto.Unmarshal(rawSBlock, &sBlock); err != nil {
+			s.Logger.Warnf("fail to decode stream block: %s", err.Error())
+		}
+
+		if sBlock.Source == s.Config.ID {
+			continue
+		}
+
+		block := ToBlockAndTrace(&sBlock)
+		if sBlock.IsCache {
+			s.m.StreamRecvCounter.WithLabelValues("cache").Inc()
+			if err := s.cachePayload(ctx, ds, block); err != nil {
+				s.Logger.WithError(err).With(block).Warn("failed to cache payload")
+			}
+		} else {
+			s.m.StreamRecvCounter.WithLabelValues("store").Inc()
+			if err := s.storePayload(ctx, ds, block); err != nil {
+				s.Logger.WithError(err).With(block).Warn("failed to store payload: %s")
+			}
+		}
+	}
+
+	return ctx.Err()
+}
+
+func (s *RedisStream) RunPublisherParallel(ctx context.Context, num uint) {
+	for i := uint(0); i < num; i++ {
+		go s.RunPublisher(ctx)
+	}
+}
+
+func (s *RedisStream) RunPublisher(ctx context.Context) error {
+	for {
+		select {
+		case req := <-s.cacheRequests:
+			s.encodeAndPublish(ctx, &req, true)
+			continue
+		default:
+		}
+
+		select {
+		case req := <-s.cacheRequests:
+			s.encodeAndPublish(ctx, &req, true)
+		case req := <-s.storeRequests:
+			s.encodeAndPublish(ctx, &req, false)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+}
+
+func (s *RedisStream) PublishStoreBlock() chan structs.BlockAndTrace {
+	return s.storeRequests
+}
+
+func (s *RedisStream) PublishCacheBlock() chan structs.BlockAndTrace {
+	return s.cacheRequests
+}
+
+func (s *RedisStream) encodeAndPublish(ctx context.Context, block *structs.BlockAndTrace, isCache bool) {
+	timer1 := prometheus.NewTimer(s.m.Timing.WithLabelValues("encodeAndPublish", "encode"))
+	rawBlock, err := proto.Marshal(ToStreamBlock(block, isCache, s.Config.ID))
+	if err != nil {
+		s.Logger.Warnf("fail to encode encode and stream block: %s", err.Error())
+		timer1.ObserveDuration()
+		return
+	}
+	timer1.ObserveDuration()
+
+	timer2 := prometheus.NewTimer(s.m.Timing.WithLabelValues("encodeAndPublish", "publish"))
+	defer timer2.ObserveDuration()
+	if err := s.Pubsub.Publish(ctx, s.Config.PubsubTopic, rawBlock); err != nil {
+		s.Logger.Warnf("fail to encode encode and stream block: %s", err.Error())
+		return
+	}
+}
+
+func (s *RedisStream) cachePayload(ctx context.Context, ds Datastore, payload *structs.BlockAndTrace) error {
```

Comment:
How do you feel about moving this cache somewhere else? 
So it would become separate entity that is unrelated to both datastore and stream.

Reply:
Why would you like to do that?